draco/analysis/flagging.py

```diff
@@ -628,49 +628,60 @@ class ThresholdVisWeight(task.SingleTask):
         ----------
         timestream : `.core.container` with `weight` attribute
 
         Returns
         -------
-        timestream : same as input timestream
-            The input container with modified weights.
+        out : RFIMask container
+            RFIMask container with mask set
         """
-        from mpi4py import MPI
-
-        timestream.redistribute(["prod", "stack"])
-
-        weight = timestream.weight[:]
-
-        # Average over the frequency and time axes to get a per baseline
-        # average
-        mean_weight = weight.mean(axis=2).mean(axis=0)
-
-        # Figure out which entries to keep
-        threshold = np.maximum(
-            self.absolute_threshold, self.relative_threshold * mean_weight
-        )
-        keep = weight > threshold[np.newaxis, :, np.newaxis]
-        keep_sum = np.sum(keep)
-        keep_total = np.zeros_like(keep_sum)
-
-        timestream.comm.Allreduce(keep_sum, keep_total, op=MPI.SUM)
-        keep_frac = keep_total / float(np.prod(weight.global_shape))
+        stream.redistribute("freq")
 
-        self.log.info(
-            "%0.5f%% of data is below the weight threshold"
-            % (100.0 * (1.0 - keep_frac))
+        # Make the output container depending on what 'stream' is
+        if "ra" in stream.axes:
+            out = containers.SiderealRFIMask(axes_from=stream, attrs_from=stream)
+        elif "time" in stream.axes:
+            out = containers.RFIMask(axes_from=stream, attrs_from=stream)
+        else:
+            raise TypeError(f"Require Timestream or SiderealStream. Got {type(stream)}")
+
+        weight = stream.weight[:]
+        # Average over the baselines (stack) axis
+        mean_baseline = np.mean(weight, axis=1, keepdims=True)
+        # Cut out any values below fixed threhold. Return a np.ndarray
+        threshold = np.where(
+            mean_baseline > self.absolute_threshold, mean_baseline, np.nan
         )
+        # Average across the time (ra) axis to get per-frequency thresholds,
+        # ignoring any nans. np.nanmean will give a warning if an entire band is
+        # nan, which we expect to happen in some cases.
+        with warnings.catch_warnings():
+            warnings.filterwarnings(action="ignore", message="Mean of empty slice")
+            threshold = np.nanmean(threshold, axis=2, keepdims=True)
+        # Create a 2D baseline-independent mask.
+        mask = ~(
+            mean_baseline.local_array
+            > np.fmax(threshold * self.relative_threshold, self.absolute_threshold)
+        )[:, 0, :]
+        # Collect all parts of the mask. Method .allgather() returns a np.ndarray
+        mask = mpiarray.MPIArray.wrap(mask, axis=0).allgather()
+        if self.comm.rank == 0:
+            # Log the percent of data masked
+            drop_frac = np.sum(mask) / np.prod(mask.shape)
+            self.log.info(
+                "%0.5f%% of data is below the weight threshold" % (100.0 * (drop_frac))
+            )
```

Comment:
I think in here you should remove the `self.comm.rank == 0` conditional, and output on all ranks. We generally favour making log messages output on all ranks, and then controlling what gets put in the logs with the rank dependent level cutoffs (you'll see that in the configs in `SetMPILogging` with the `level_rank0` etc parameters).